src/cast_core/src/actors/CrudActor.cpp

```diff
@@ -0,0 +1,1241 @@
+        }
+        return true;
+    }
+};
+}  // namespace YAML
+
+namespace genny::actor {
+
+namespace {
+
+struct BaseOperation {
+    virtual void run(mongocxx::client_session& session) = 0;
+    virtual ~BaseOperation() = default;
+};
+
+using OpCallback = std::function<std::unique_ptr<BaseOperation>(
+    YAML::Node, bool, mongocxx::collection, genny::DefaultRandom&, metrics::Operation)>;
+
+struct WriteOperation : public BaseOperation {
+    virtual mongocxx::model::write getModel() = 0;
+};
+
+using WriteOpCallback = std::function<std::unique_ptr<WriteOperation>(
+    YAML::Node, bool, mongocxx::collection, genny::DefaultRandom&, metrics::Operation)>;
+
+struct InsertOneOperation : public WriteOperation {
+    InsertOneOperation(YAML::Node opNode,
+                       bool onSession,
+                       mongocxx::collection collection,
+                       genny::DefaultRandom& rng,
+                       metrics::Operation operation)
+        : _onSession{onSession}, _collection{collection}, _operation{operation} {
+        auto insertDoc = opNode["Document"];
+        if (!insertDoc) {
+            throw InvalidConfigurationException("'insertOne' expects a 'Document' field.");
+        }
+        _docTemplate = value_generators::makeDoc(insertDoc, rng);
+
+        // TODO: parse insert options.
+    }
+
+    mongocxx::model::write getModel() override {
+        bsoncxx::builder::stream::document document{};
+        _docTemplate->view(document);
+        return mongocxx::model::insert_one{document.extract()};
+    }
+
+    void run(mongocxx::client_session& session) override {
+        bsoncxx::builder::stream::document insertDoc{};
+        _docTemplate->view(insertDoc);
+        auto view = insertDoc.view();
+        auto ctx = _operation.start();
+        (_onSession) ? _collection.insert_one(session, view, _options)
+                     : _collection.insert_one(view, _options);
+        ctx.success();
+    }
+
+private:
+    bool _onSession;
+    mongocxx::collection _collection;
+    std::unique_ptr<DocGenerator> _docTemplate;
+    metrics::Operation _operation;
+    mongocxx::options::insert _options;
+};
+
+struct UpdateOneOperation : public WriteOperation {
+    UpdateOneOperation(YAML::Node opNode,
+                       bool onSession,
+                       mongocxx::collection collection,
+                       genny::DefaultRandom& rng,
+                       metrics::Operation operation)
+        : _onSession{onSession}, _collection{collection}, _operation{operation} {
+        auto filter = opNode["Filter"];
+        auto update = opNode["Update"];
+        if (!filter || !update) {
+            throw InvalidConfigurationException(
+                "'updateOne' expects 'Filter' and 'Update' fields.");
+        }
+        _filterTemplate = value_generators::makeDoc(filter, rng);
+        _updateTemplate = value_generators::makeDoc(update, rng);
+
+        // TODO: parse update options.
```

Comment:
not sure what this and the other TODO means - can you elaborate? Seems like it's already parsed?

Reply:
Even better, create a Jira ticket for all the TODOs in this PR